extensions/wikidata/src/org/openrefine/wikidata/editing/EditBatchProcessor.java

```diff
@@ -0,0 +1,210 @@
+
+/**
+ * Schedules and performs a list of updates to items via the API.
+ * 
+ * @author Antonin Delpeuch
+ *
+ */
+public class EditBatchProcessor {
+
+    static final Logger logger = LoggerFactory.getLogger(EditBatchProcessor.class);
+
+    private WikibaseDataFetcher fetcher;
+    private WikibaseDataEditor editor;
+    private NewItemLibrary library;
+    private List<ItemUpdate> scheduled;
+    private String summary;
+
+    private List<ItemUpdate> remainingUpdates;
+    private List<ItemUpdate> currentBatch;
+    private int batchCursor;
+    private int globalCursor;
+    private Map<String, EntityDocument> currentDocs;
+    private int batchSize;
+
+    /**
+     * Initiates the process of pushing a batch of updates to Wikibase. This
+     * schedules the updates and is a prerequisite for calling
+     * {@link performOneEdit}.
+     * 
+     * @param fetcher
+     *            the fetcher to use to retrieve the current state of items
+     * @param editor
+     *            the object to use to perform the edits
+     * @param updates
+     *            the list of item updates to perform
+     * @param library
+     *            the library to use to keep track of new item creation
+     * @param summary
+     *            the summary to append to all edits
+     * @param batchSize
+     *            the number of items that should be retrieved in one go from the
+     *            API
+     */
+    public EditBatchProcessor(WikibaseDataFetcher fetcher, WikibaseDataEditor editor, List<ItemUpdate> updates,
+            NewItemLibrary library, String summary, int batchSize) {
+        this.fetcher = fetcher;
+        this.editor = editor;
+        editor.setEditAsBot(true); // this will not do anything if the user does not
+        // have a bot flag, and this is generally wanted if they have one.
+        this.library = library;
+        this.summary = summary;
+        this.batchSize = batchSize;
+
+        // Schedule the edit batch
+        WikibaseAPIUpdateScheduler scheduler = new WikibaseAPIUpdateScheduler();
+        this.scheduled = scheduler.schedule(updates);
+        this.globalCursor = 0;
+
+        this.batchCursor = 0;
+        this.remainingUpdates = new ArrayList<>(scheduled);
+        this.currentBatch = Collections.emptyList();
+        this.currentDocs = Collections.emptyMap();
+    }
+
+    /**
+     * Performs the next edit in the batch.
+     * 
+     * @throws InterruptedException
+     */
+    public void performEdit()
+            throws InterruptedException {
+        if (remainingEdits() == 0) {
+            return;
+        }
+        if (batchCursor == currentBatch.size()) {
+            prepareNewBatch();
+        }
+        ItemUpdate update = currentBatch.get(batchCursor);
+
+        // Rewrite mentions to new items
+        ReconEntityRewriter rewriter = new ReconEntityRewriter(library, update.getItemId());
+        update = rewriter.rewrite(update);
```

Comment:
A little bit weird to pass "update" and return "update". Can skip the return?

Reply:
what do you mean?

Reply:
I mean:
rewriter.rewrite(update);

Reply:
@jackyq2015 it was like this before, but I have changed EntityUpdate to be immutable (just like the other classes in Wikidata-Toolkit). It's something we don't do much in OpenRefine (everything is public and mutable) but it brings a lot of good things (thread safety, easier testing)…